mlflow/pyspark/ml/__init__.py

```diff
@@ -0,0 +1,282 @@
+    """
+    allowlist_file = spark_session.sparkContext._conf.get(
+        "spark.databricks.mlflow.autolog.log_model.allowlistFile", None
+    )
+    builtin_allowlist_file = resource_filename(__name__, "log_model_allowlist.txt")
+    if allowlist_file:
+        try:
+            return _read_log_model_allowlist_from_file(allowlist_file)
+        except Exception:
+            # fallback to built-in allowlist file
+            _logger.warning(
+                "Reading config spark.databricks.mlflow.autolog.log_model.allowlistFile "
+                f"{allowlist_file} failed, fallback to built-in allowlist file"
+            )
+            return _read_log_model_allowlist_from_file(builtin_allowlist_file)
+    else:
+        return _read_log_model_allowlist_from_file(builtin_allowlist_file)
+
+
+# When autolog called, `_log_model_allowlist` will be initialized.
+_log_model_allowlist = None
+
+
+def _get_warning_msg_for_skip_log_model(model):
+    return (
+        f"This model {model.uid} is not logged because it is not in "
+        "allowlist or its nested models are not in allowlist."
+    )
+
+
+def _should_log_model(spark_model):
+    # TODO: Handle PipelineModel/CrossValidatorModel/TrainValidationSplitModel
+    class_name = _get_fully_qualified_class_name(spark_model)
+    if class_name in _log_model_allowlist:
+        if class_name == "pyspark.ml.classification.OneVsRestModel":
+            return _should_log_model(spark_model.models[0])
+        else:
+            return True
+    else:
+        return False
+
+
+def _get_estimator_info_tags(estimator):
+    """
+    :return: A dictionary of MLflow run tag keys and values
+             describing the specified estimator.
+    """
+    return {
+        "estimator_name": estimator.__class__.__name__,
+        "estimator_class": _get_fully_qualified_class_name(estimator),
+    }
+
+
+def _get_instance_param_map(instance):
+    from pyspark.ml.param import Params
+
+    param_map = {
+        param.name: instance.getOrDefault(param)
+        for param in instance.params
+        if instance.isDefined(param)
+    }
+    # TODO:
+    #  handle special case: CrossValidator.estimatorParamMaps
+
+    expanded_param_map = {}
+    for k, v in param_map.items():
+        if isinstance(v, Params):
+            expanded_param_map[k] = v.uid
+            internal_param_map = _get_instance_param_map(v)
+            for ik, iv in internal_param_map.items():
+                expanded_param_map[f"{v.uid}.{ik}"] = iv
+        else:
+            expanded_param_map[k] = v
+
+    return expanded_param_map
+
+
+# NOTE: The current implementation doesn't guarantee thread-safety, but that's okay for now because:
+# 1. We don't currently have any use cases for allow_children=True.
+# 2. The list append & pop operations are thread-safe, so we will always clear the session stack
+#    once all _SparkTrainingSessions exit.
+class _SparkTrainingSession(object):
```

Comment:
Do we need to define a new `__SparkTrainingSession` helper class for spark, or could we reuse the one from sklearn autologging?

Reply:
We could re-use. I think we can do refactor in follow-up PR.